pkg/frontend/querymiddleware/querysharding_test.go

```diff
@@ -1635,48 +1637,79 @@ func TestQuerySharding_ShouldReturnErrorInCorrectFormat(t *testing.T) {
 	)
 
 	for _, tc := range []struct {
-		name             string
-		engineDownstream *promql.Engine
-		engineSharding   *promql.Engine
-		expError         error
-		queryable        storage.Queryable
+		name                 string
+		engineType           string
+		engineDownstreamOpts []engineOpt
+		engineShardingOpts   []engineOpt
+		expError             error
+		queryable            storage.Queryable
 	}{
+		// Prometheus engine tests.
+
+		{
+			name:                 "downstream - timeout",
+			engineType:           querier.PrometheusEngine,
+			engineDownstreamOpts: []engineOpt{withTimeout(50 * time.Millisecond)},
+			expError:             apierror.New(apierror.TypeTimeout, "query timed out in expression evaluation"),
+			queryable:            queryableSlow,
+		},
+		{
+			name:                 "downstream - sample limit",
+			engineType:           querier.PrometheusEngine,
+			engineDownstreamOpts: []engineOpt{withMaxSamples(1)},
+			expError:             apierror.New(apierror.TypeExec, "query processing would load too many samples into memory in query execution"),
+		},
 		{
-			name:             "downstream - timeout",
-			engineDownstream: engineTimeout,
-			engineSharding:   engine,
-			expError:         apierror.New(apierror.TypeTimeout, "query timed out in expression evaluation"),
-			queryable:        queryableSlow,
+			name:               "sharding - timeout",
+			engineType:         querier.PrometheusEngine,
+			engineShardingOpts: []engineOpt{withTimeout(50 * time.Millisecond)},
+			expError:           apierror.New(apierror.TypeTimeout, "query timed out in expression evaluation"),
+			queryable:          queryableSlow,
 		},
 		{
-			name:             "sharding - sample limit",
-			engineDownstream: engineSampleLimit,
-			engineSharding:   engine,
-			expError:         apierror.New(apierror.TypeExec, "query processing would load too many samples into memory in query execution"),
+			name:       "downstream - storage internal error",
+			engineType: querier.PrometheusEngine,
+			queryable:  queryableInternalErr,
+			expError:   apierror.New(apierror.TypeInternal, "some internal error"),
 		},
 		{
-			name:             "sharding - timeout",
-			engineDownstream: engine,
-			engineSharding:   engineTimeout,
-			expError:         apierror.New(apierror.TypeTimeout, "query timed out in expression evaluation"),
-			queryable:        queryableSlow,
+			name:       "downstream - storage prometheus execution error",
+			engineType: querier.PrometheusEngine,
+			queryable:  queryablePrometheusExecErr,
+			expError:   apierror.Newf(apierror.TypeExec, "expanding series: %s", querier.NewMaxQueryLengthError(744*time.Hour, 720*time.Hour)),
+		},
+
+		// MQE equivalents when applicable. For example, MQE doesn't have a sample limit
+		// and returns different error messages for timeouts when executing a query.
```

Comment:
Rather than duplicating the tests like this, what if we added a flag to disable the test cases that aren't relevant for MQE, and to provide alternative error messages if MQE is being used?

Reply:
I've added a flag to skip MQE for the sample test and stop comparing the bodies generated by the errors (this was not a great idea anyway) and instead now just compare the error types and HTTP status codes.